pkg/ext/resources/tokens/store.go

```diff
@@ -0,0 +1,325 @@
+package tokens
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rancher/rancher/pkg/auth/tokens/hashers"
+	"github.com/rancher/rancher/pkg/ext/resources/types"
+	v1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/v3/pkg/randomtoken"
+	authv1 "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	authzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const TokenNamespace = "cattle-token-data"
+
+// +k8s:openapi-gen=false
+// +k8s:deepcopy-gen=false
+type TokenStore struct {
+	secretClient v1.SecretClient
+	secretCache  v1.SecretCache
+	sar          authzv1.SubjectAccessReviewInterface
+}
+
+func NewTokenStore(secretClient v1.SecretClient, secretCache v1.SecretCache, sar authzv1.SubjectAccessReviewInterface) types.Store[*RancherToken, *RancherTokenList] {
+	tokenStore := TokenStore{
+		secretClient: secretClient,
+		secretCache:  secretCache,
+		sar:          sar,
+	}
+	return &tokenStore
+}
+
+func (t *TokenStore) Create(ctx context.Context, userInfo user.Info, token *RancherToken, opts *metav1.CreateOptions) (*RancherToken, error) {
+	if token.Status.PlaintextToken == "" {
+		tokenValue, err := randomtoken.Generate()
+		if err != nil {
+			return nil, fmt.Errorf("unable to generate token value: %w", err)
+		}
+		token.Status.PlaintextToken = tokenValue
+		hashedValue, err := hashers.GetHasher().CreateHash(tokenValue)
+		if err != nil {
+			return nil, fmt.Errorf("unable to hash token value: %w", err)
+		}
+		token.Status.HashedToken = hashedValue
+	}
```

Comment:
The entire block implies that the user can supply an arbitrary token value.
Do we want to allow that ?
IIRC that was one of the disliked/rejected things of the pre-imp-API token rfc.
New spec currently does not say one way or other.
New spec example is for random generated token value.

IMHO we should disallow and write this into the spec.
In that case `PlaintextToken != ""` check should trigger an error here.